modules/core/04-channel/keeper/upgrade_test.go

```diff
@@ -126,6 +129,185 @@ func (suite *KeeperTestSuite) TestChanUpgradeInit() {
+			func() {
+				channel := path.EndpointB.GetChannel()
+				channel.UpgradeSequence++
+				path.EndpointB.SetChannel(channel)
+			},
+			types.ErrIncompatibleCounterpartyUpgrade,
+		},
+		{
+			"upgrade ordering is not the same on both sides",
+			func() {
+				upgrade.Fields.Ordering = types.ORDERED
+			},
+			types.ErrIncompatibleCounterpartyUpgrade,
+		},
+		{
+			"proposed connection is not found",
+			func() {
+				upgrade.Fields.ConnectionHops[0] = "invalid-connection"
+			},
+			connectiontypes.ErrConnectionNotFound,
+		},
+		{
+			"proposed connection is not in OPEN state",
+			func() {
+				// reuse existing connection to create a new connection in a non OPEN state
+				connectionEnd := path.EndpointB.GetConnection()
+				connectionEnd.State = connectiontypes.UNINITIALIZED
+				connectionEnd.Counterparty.ConnectionId = counterpartyUpgrade.Fields.ConnectionHops[0] // both sides must be each other's counterparty
+
+				// set proposed connection in state
+				proposedConnectionID := "connection-100"
+				suite.chainB.GetSimApp().GetIBCKeeper().ConnectionKeeper.SetConnection(suite.chainB.GetContext(), proposedConnectionID, connectionEnd)
+				upgrade.Fields.ConnectionHops[0] = proposedConnectionID
+			},
+			connectiontypes.ErrInvalidConnectionState,
+		},
+		{
+			"proposed connection ends are not each other's counterparty",
+			func() {
+				// reuse existing connection to create a new connection in a non OPEN state
+				connectionEnd := path.EndpointB.GetConnection()
+				// ensure counterparty connectionID does not match connectionID set in counterparty proposed upgrade
+				connectionEnd.Counterparty.ConnectionId = "connection-50"
+
+				// set proposed connection in state
+				proposedConnectionID := "connection-100"
+				suite.chainB.GetSimApp().GetIBCKeeper().ConnectionKeeper.SetConnection(suite.chainB.GetContext(), proposedConnectionID, connectionEnd)
+				upgrade.Fields.ConnectionHops[0] = proposedConnectionID
+			},
+			types.ErrIncompatibleCounterpartyUpgrade,
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		suite.Run(tc.name, func() {
+			suite.SetupTest()
+
+			path = ibctesting.NewPath(suite.chainA, suite.chainB)
+			suite.coordinator.Setup(path)
+
+			upgradeVersion := fmt.Sprintf("%s-v2", mock.Version)
+			path.EndpointA.ChannelConfig.ProposedUpgrade.Fields.Version = upgradeVersion
+			err := path.EndpointA.ChanUpgradeInit()
+			suite.Require().NoError(err)
+
+			// ensure proof verification succeeds
+			err = path.EndpointB.UpdateClient()
+			suite.Require().NoError(err)
+
+			// same construction used by endpoint.ChanUpgradeTry
+			channelKey := host.ChannelKey(path.EndpointA.ChannelConfig.PortID, path.EndpointA.ChannelID)
+			proofChannel, proofHeight := path.EndpointA.Chain.QueryProof(channelKey)
+			upgradeKey := host.ChannelUpgradeKey(path.EndpointA.ChannelConfig.PortID, path.EndpointA.ChannelID)
+			proofUpgrade, _ := path.EndpointA.Chain.QueryProof(upgradeKey)
```

Comment:
#3722

Reply:
Nice, agree this would improve readability and clutter of the test body :)